making some change on Daikin176bits to work with IRMQTTServer

pasna opened the pull request:
Most function works except fanspeed.

pasna committed:
Add files via upload

pasna committed:
add Daikin176bit protocol

pasna committed:
Multiple corrections of white space

pasna committed:
Daikin176 listing correction

pasna committed:
correction on indent

pasna committed:
minor correction on whitespace

pasna committed:
Update ir_Daikin.cpp

pasna committed:
Delete .DS_Store

pasna committed:
Delete .DS_Store

pasna committed:
minor correction on indents

pasna committed:
more indents correction

pasna committed:
further correction on indents

pasna committed:
more correction on line end  whitespace

pasna committed:
one more extra whitespace remove

hopefully this is the last whitespace .

pasna committed:
Merge pull request #3 from pasna/Daikin176bits

Daikin176bits

pasna committed:
update

pasna committed:
update Daikin176bits

pasna committed:
Rename IRMQTTServer/IRMQTTServer.h to example/IRMQTTServer/IRMQTTServer.h

pasna committed:
Delete IRMQTTServer.h

pasna committed:
Delete IRMQTTServer.h

pasna committed:
Delete platformio.ini

pasna committed:
Rename IRMQTTServer/platformio.ini to examples/IRMQTTServer/platformio.ini

pasna committed:
Add files via upload

pasna committed:
Delete IRrecvDumpV2.ino

pasna committed:
Delete platformio.ini

pasna committed:
Rename IRrecvDumpV2/IRrecvDumpV2.ino to examples/IRrecvDumpV2/IRrecvDumpV2.ino

pasna committed:
Rename IRrecvDumpV2/platformio.ini to examples/IRrecvDumpV2/platformio.ini

pasna committed:
Delete IRMQTTServer.h

pasna committed:
Delete IRMQTTServer.ino

pasna committed:
Delete platformio.ini

pasna committed:
Delete IRrecvDumpV2.ino

pasna committed:
Delete platformio.ini

pasna committed:
Add files via upload

pasna committed:
Add files via upload

pasna committed:
Add files via upload

pasna committed:
Merge pull request #6 from crankyoldgit/master

update to latest

pasna committed:
Add files via upload

pasna committed:
Merge pull request #7 from pasna/pasna-patch-1

add Daikin 176bits to IRMQTTServer

pasna committed:
fixed few bugs 

fixed few bugs error compiling IRMQTTServer

pasna committed:
Merge pull request #9 from pasna/pasna-patch-5

fixed few bugs

pasna committed:
fixed bugs Daikin176 for IRMQTTServer

pasna committed:
Merge pull request #10 from pasna/fixed-bugDaikin176-for-IRMQTTServer-

fixed bugs Daikin176 for IRMQTTServer

pasna commented:
Hi @crankyoldgit 
Please help fixing fanspeed function, I tried but couldn't figure out.
Thanks.

crankyoldgit commented:
@pasna Sure.
What are the data values you've recorded for the fan settings for this model, and what are the corresponding values/buttons on the remote?

pasna commented:
@crankyoldgit Fan setting shares byte 18 (left digit) to horizontal swing (right digit). Fan has 2 speeds only, 0x1 is min and 0x3 is max. Horizontal setting has 2 settings 0x05 is ON and 0x06 is OFF.
Thank you and best regards.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -2310,7 +2312,7 @@ void IRDaikin176::setFan(const uint8_t fan) {
   else if (fan < kDaikinFanMin || fan > kDaikinFanMax)
```
`kDaikinFanMax` is `5`

Your comment says it's 3. Thus you can't use that same constant. You need a new one, if your comment is correct. e.g. 3.

Try:
`const uint8_t kDaikin176FanMax = 3;`

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -2323,12 +2325,12 @@ uint8_t IRDaikin176::getFan() {
-    case stdAc::fanspeed_t::kMedium: return kDaikinFanMin + 2;
-    case stdAc::fanspeed_t::kHigh: return kDaikinFanMax - 1;
-    case stdAc::fanspeed_t::kMax: return kDaikinFanMax;
+    case stdAc::fanspeed_t::kLow:
```
This bit looks wrong to me.
You are returning `kDaikinFanMax - 2` for all of Low, Medium, High, & Max. 

You possibly want:
```
switch (speed) {
  case stdAc::fanspeed_t::kMin:
  case stdAc::fanspeed_t::kLow:
    return kDaikinFanMin;
  case stdAc::fanspeed_t::kMedium:
    return kDaikinFanMin + 1;
  case stdAc::fanspeed_t::kHigh:
  case stdAc::fanspeed_t::kMax:
    return kDaikin176FanMax;
  ...
```

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -2360,7 +2366,21 @@ uint8_t IRDaikin176::convertSwingH(const stdAc::swingh_t position) {
+
+// Convert a native fan speed to it's common equivalent.
+stdAc::fanspeed_t IRDaikin176::toCommonFanSpeed(const uint8_t speed) {
+  switch (speed) {
```
Again, similarly wrong. Try:
``` 
 switch (speed) {
    case kDaikin176FanMax: return stdAc::fanspeed_t::kMax;
    case kDaikinFanMin + 1: return stdAc::fanspeed_t::kMedium;
    case kDaikinFanQuiet:
    case kDaikinFanMin: return stdAc::fanspeed_t::kMin;
    default: return stdAc::fanspeed_t::kAuto;
  }
```

crankyoldgit reviewed (changes requested):
Have a look at those comments and see if that fixes things.
The bit shifting & masking etc looks fine to me for the set/getFan() methods.

pasna committed:
modified fan setting Daikin176bits to match remote

pasna commented:
Hi @crankyoldgit 
I had modified according to you suggestion, unfortunately fan setting still unable to transmit by IRMQTTServer. Please see below.

transmittion from IRMQTTServer capture with IRrecvDumpV2 - (byte 18):


Timestamp : 000581.729
Library   : v2.6.3

Encoding  : DAIKIN176
Code      : 11DA171804001E11DA171800730021000022050020F5 (176 bits)
Mesg Desc.: Power: On, Mode: 7 (COOL), Temp: 26C, Fan: 0, Swing (H): 5 (Auto)
Raw Timing[359]:
   +  5060, -  2170,    +   366, -  1790,    +   364, -   718,    +   366, -   720, 
   +   364, -   718,    +   366, -  1790,    +   364, -   720,    +   364, -   718, 
   +   366, -   720,    +   366, -   720,    +   364, -  1790,    +   366, -   718, 
   +   366, -  1790,    +   364, -  1790,    +   364, -   720,    +   364, -  1790, 
   +   364, -  1790,    +   366, -  1790,    +   364, -  1788,    +   376, -  1784, 
   +   364, -   720,    +   366, -  1788,    +   366, -   720,    +   364, -   722, 
   +   364, -   720,    +   366, -   718,    +   366, -   718,    +   366, -   720, 
   +   364, -  1790,    +   366, -  1788,    +   366, -   720,    +   364, -   720, 
   +   366, -   720,    +   364, -   720,    +   364, -   720,    +   366, -  1788, 
   +   366, -   722,    +   364, -   718,    +   366, -   720,    +   364, -   720, 
   +   364, -   720,    +   366, -   718,    +   364, -   720,    +   364, -   720, 
   +   364, -   720,    +   364, -   720,    +   364, -   746,    +   338, -   720, 
   +   364, -   720,    +   366, -   720,    +   364, -  1790,    +   366, -  1790, 
   +   366, -  1790,    +   364, -  1788,    +   366, -   720,    +   366, -   720, 
   +   364, -   720,    +   366, - 29490,    +  5094, -  2144,    +   368, -  1790, 
   +   364, -   720,    +   366, -   720,    +   364, -   720,    +   364, -  1790, 
   +   366, -   718,    +   366, -   722,    +   364, -   720,    +   364, -   720, 
   +   364, -  1790,    +   366, -   744,    +   340, -  1790,    +   364, -  1788, 
   +   366, -   718,    +   366, -  1790,    +   364, -  1790,    +   366, -  1788, 
   +   366, -  1790,    +   364, -  1792,    +   364, -   720,    +   366, -  1790, 
   +   364, -   718,    +   366, -   720,    +   364, -   746,    +   340, -   720, 
   +   364, -   720,    +   366, -   720,    +   364, -  1790,    +   364, -  1788, 
   +   366, -   720,    +   364, -   720,    +   366, -   746,    +   340, -   718, 
   +   366, -   720,    +   366, -   720,    +   364, -   718,    +   366, -   720, 
   +   364, -   718,    +   366, -   720,    +   366, -   720,    +   366, -  1788, 
   +   366, -  1790,    +   366, -   718,    +   366, -   720,    +   364, -  1790, 
   +   364, -  1790,    +   364, -  1790,    +   364, -   720,    +   364, -   720, 
   +   364, -   720,    +   364, -   720,    +   366, -   718,    +   366, -   718, 
   +   366, -   720,    +   366, -   720,    +   364, -   718,    +   366, -  1790, 
   +   364, -   720,    +   366, -   720,    +   364, -   720,    +   366, -   718, 
   +   366, -  1790,    +   366, -   744,    +   340, -   720,    +   364, -   720, 
   +   364, -   720,    +   366, -   720,    +   364, -   720,    +   364, -   720, 
   +   364, -   746,    +   338, -   720,    +   366, -   720,    +   364, -   722, 
   +   362, -   720,    +   370, -   720,    +   364, -   720,    +   366, -   720, 
   +   368, -   720,    +   364, -   746,    +   338, -   720,    +   366, -   720, 
   +   364, -  1790,    +   364, -   718,    +   366, -   720,    +   364, -   720, 
   +   364, -  1790,    +   364, -   720,    +   364, -   720,    +   366, -  1788, 
   +   366, -   720,    +   366, -  1788,    +   366, -   720,    +   364, -   720, 
   +   364, -   720,    +   364, -   720,    +   364, -   720,    +   366, -   722, 
   +   364, -   718,    +   366, -   718,    +   366, -   720,    +   364, -   720, 
   +   366, -   718,    +   366, -   746,    +   338, -   722,    +   364, -   720, 
   +   364, -   718,    +   366, -   722,    +   362, -   720,    +   364, -   720, 
   +   366, -  1788,    +   366, -   720,    +   362, -   722,    +   364, -  1792, 
   +   364, -   720,    +   364, -  1790,    +   364, -   720,    +   364, -  1790, 
   +   366, -  1790,    +   364, -  1774,    +   382, -  1790,    +   366

uint16_t rawData[359] = {5060, 2170,  366, 1790,  364, 718,  366, 720,  364, 718,  366, 1790,  364, 720,  364, 718,  366, 720,  366, 720,  364, 1790,  366, 718,  366, 1790,  364, 1790,  364, 720,  364, 1790,  364, 1790,  366, 1790,  364, 1788,  376, 1784,  364, 720,  366, 1788,  366, 720,  364, 722,  364, 720,  366, 718,  366, 718,  366, 720,  364, 1790,  366, 1788,  366, 720,  364, 720,  366, 720,  364, 720,  364, 720,  366, 1788,  366, 722,  364, 718,  366, 720,  364, 720,  364, 720,  366, 718,  364, 720,  364, 720,  364, 720,  364, 720,  364, 746,  338, 720,  364, 720,  366, 720,  364, 1790,  366, 1790,  366, 1790,  364, 1788,  366, 720,  366, 720,  364, 720,  366, 29490,  5094, 2144,  368, 1790,  364, 720,  366, 720,  364, 720,  364, 1790,  366, 718,  366, 722,  364, 720,  364, 720,  364, 1790,  366, 744,  340, 1790,  364, 1788,  366, 718,  366, 1790,  364, 1790,  366, 1788,  366, 1790,  364, 1792,  364, 720,  366, 1790,  364, 718,  366, 720,  364, 746,  340, 720,  364, 720,  366, 720,  364, 1790,  364, 1788,  366, 720,  364, 720,  366, 746,  340, 718,  366, 720,  366, 720,  364, 718,  366, 720,  364, 718,  366, 720,  366, 720,  366, 1788,  366, 1790,  366, 718,  366, 720,  364, 1790,  364, 1790,  364, 1790,  364, 720,  364, 720,  364, 720,  364, 720,  366, 718,  366, 718,  366, 720,  366, 720,  364, 718,  366, 1790,  364, 720,  366, 720,  364, 720,  366, 718,  366, 1790,  366, 744,  340, 720,  364, 720,  364, 720,  366, 720,  364, 720,  364, 720,  364, 746,  338, 720,  366, 720,  364, 722,  362, 720,  370, 720,  364, 720,  366, 720,  368, 720,  364, 746,  338, 720,  366, 720,  364, 1790,  364, 718,  366, 720,  364, 720,  364, 1790,  364, 720,  364, 720,  366, 1788,  366, 720,  366, 1788,  366, 720,  364, 720,  364, 720,  364, 720,  364, 720,  366, 722,  364, 718,  366, 718,  366, 720,  364, 720,  366, 718,  366, 746,  338, 722,  364, 720,  364, 718,  366, 722,  362, 720,  364, 720,  366, 1788,  366, 720,  362, 722,  364, 1792,  364, 720,  364, 1790,  364, 720,  364, 1790,  366, 1790,  364, 1774,  382, 1790,  366};  // DAIKIN176
uint8_t state[22] = {0x11, 0xDA, 0x17, 0x18, 0x04, 0x00, 0x1E, 0x11, 0xDA, 0x17, 0x18, 0x00, 0x73, 0x00, 0x21, 0x00, 0x00, 0x22, 0x05, 0x00, 0x20, 0xF5};

while transmit from actual A/C remote 
Timestamp : 000593.021
Library   : v2.6.3

Encoding  : DAIKIN176
Code      : 11DA171804001E11DA17180073002100002235002025 (176 bits)
Mesg Desc.: Power: On, Mode: 7 (COOL), Temp: 26C, Fan: 3 (MAX), Swing (H): 5 (Auto)
Raw Timing[359]:
   +  5084, -  2136,    +   384, -  1804,    +   356, -   696,    +   382, -   724, 
   +   356, -   696,    +   382, -  1778,    +   382, -   724,    +   356, -   724, 
   +   356, -   698,    +   382, -   696,    +   382, -  1778,    +   382, -   724, 
   +   356, -  1776,    +   384, -  1776,    +   384, -   722,    +   356, -  1776, 
   +   384, -  1778,    +   382, -  1804,    +   356, -  1804,    +   356, -  1778, 
   +   382, -   696,    +   384, -  1778,    +   382, -   724,    +   356, -   724, 
   +   356, -   724,    +   356, -   696,    +   384, -   724,    +   356, -   698, 
   +   382, -  1778,    +   382, -  1776,    +   384, -   724,    +   356, -   698, 
   +   382, -   696,    +   384, -   724,    +   356, -   696,    +   382, -  1804, 
   +   356, -   696,    +   382, -   724,    +   356, -   698,    +   380, -   698, 
   +   382, -   698,    +   382, -   724,    +   356, -   724,    +   356, -   724, 
   +   356, -   724,    +   356, -   724,    +   356, -   724,    +   356, -   696, 
   +   382, -   724,    +   356, -   698,    +   382, -  1804,    +   356, -  1778, 
   +   382, -  1776,    +   384, -  1804,    +   356, -   724,    +   356, -   724, 
   +   356, -   698,    +   382, - 29402,    +  5084, -  2138,    +   382, -  1804, 
   +   356, -   696,    +   384, -   724,    +   356, -   698,    +   382, -  1778, 
   +   380, -   698,    +   384, -   696,    +   384, -   698,    +   382, -   698, 
   +   382, -  1780,    +   380, -   698,    +   382, -  1778,    +   382, -  1804, 
   +   356, -   678,    +   400, -  1778,    +   382, -  1804,    +   356, -  1778, 
   +   382, -  1778,    +   382, -  1778,    +   382, -   698,    +   382, -  1776, 
   +   384, -   724,    +   356, -   724,    +   356, -   724,    +   356, -   698, 
   +   382, -   724,    +   356, -   698,    +   382, -  1776,    +   384, -  1778, 
   +   382, -   724,    +   356, -   722,    +   356, -   724,    +   356, -   698, 
   +   380, -   698,    +   382, -   698,    +   382, -   724,    +   356, -   724, 
   +   356, -   724,    +   356, -   724,    +   354, -   724,    +   354, -  1778, 
   +   382, -  1778,    +   382, -   724,    +   356, -   698,    +   382, -  1760, 
   +   400, -  1778,    +   382, -  1804,    +   356, -   696,    +   382, -   696, 
   +   384, -   724,    +   354, -   724,    +   356, -   724,    +   356, -   698, 
   +   382, -   696,    +   382, -   696,    +   382, -   724,    +   356, -  1804, 
   +   356, -   724,    +   354, -   724,    +   356, -   698,    +   380, -   724, 
   +   356, -  1778,    +   382, -   724,    +   356, -   698,    +   382, -   696, 
   +   382, -   698,    +   382, -   724,    +   356, -   696,    +   382, -   724, 
   +   356, -   698,    +   380, -   724,    +   356, -   696,    +   382, -   696, 
   +   384, -   724,    +   356, -   696,    +   382, -   724,    +   356, -   724, 
   +   356, -   696,    +   384, -   696,    +   384, -   722,    +   356, -   698, 
   +   382, -  1806,    +   356, -   724,    +   356, -   698,    +   382, -   696, 
   +   384, -  1778,    +   384, -   696,    +   382, -   698,    +   382, -  1778, 
   +   382, -   724,    +   354, -  1776,    +   384, -   724,    +   356, -  1776, 
   +   384, -  1778,    +   382, -   698,    +   380, -   724,    +   356, -   696, 
   +   382, -   724,    +   356, -   724,    +   356, -   724,    +   354, -   724, 
   +   354, -   724,    +   356, -   724,    +   356, -   698,    +   382, -   724, 
   +   356, -   724,    +   356, -   696,    +   384, -   696,    +   384, -   724, 
   +   356, -  1778,    +   382, -   724,    +   356, -   696,    +   382, -  1776, 
   +   384, -   724,    +   356, -  1804,    +   356, -   696,    +   384, -   724, 
   +   356, -  1776,    +   384, -   724,    +   356, -   722,    +   358

uint16_t rawData[359] = {5084, 2136,  384, 1804,  356, 696,  382, 724,  356, 696,  382, 1778,  382, 724,  356, 724,  356, 698,  382, 696,  382, 1778,  382, 724,  356, 1776,  384, 1776,  384, 722,  356, 1776,  384, 1778,  382, 1804,  356, 1804,  356, 1778,  382, 696,  384, 1778,  382, 724,  356, 724,  356, 724,  356, 696,  384, 724,  356, 698,  382, 1778,  382, 1776,  384, 724,  356, 698,  382, 696,  384, 724,  356, 696,  382, 1804,  356, 696,  382, 724,  356, 698,  380, 698,  382, 698,  382, 724,  356, 724,  356, 724,  356, 724,  356, 724,  356, 724,  356, 696,  382, 724,  356, 698,  382, 1804,  356, 1778,  382, 1776,  384, 1804,  356, 724,  356, 724,  356, 698,  382, 29402,  5084, 2138,  382, 1804,  356, 696,  384, 724,  356, 698,  382, 1778,  380, 698,  384, 696,  384, 698,  382, 698,  382, 1780,  380, 698,  382, 1778,  382, 1804,  356, 678,  400, 1778,  382, 1804,  356, 1778,  382, 1778,  382, 1778,  382, 698,  382, 1776,  384, 724,  356, 724,  356, 724,  356, 698,  382, 724,  356, 698,  382, 1776,  384, 1778,  382, 724,  356, 722,  356, 724,  356, 698,  380, 698,  382, 698,  382, 724,  356, 724,  356, 724,  356, 724,  354, 724,  354, 1778,  382, 1778,  382, 724,  356, 698,  382, 1760,  400, 1778,  382, 1804,  356, 696,  382, 696,  384, 724,  354, 724,  356, 724,  356, 698,  382, 696,  382, 696,  382, 724,  356, 1804,  356, 724,  354, 724,  356, 698,  380, 724,  356, 1778,  382, 724,  356, 698,  382, 696,  382, 698,  382, 724,  356, 696,  382, 724,  356, 698,  380, 724,  356, 696,  382, 696,  384, 724,  356, 696,  382, 724,  356, 724,  356, 696,  384, 696,  384, 722,  356, 698,  382, 1806,  356, 724,  356, 698,  382, 696,  384, 1778,  384, 696,  382, 698,  382, 1778,  382, 724,  354, 1776,  384, 724,  356, 1776,  384, 1778,  382, 698,  380, 724,  356, 696,  382, 724,  356, 724,  356, 724,  354, 724,  354, 724,  356, 724,  356, 698,  382, 724,  356, 724,  356, 696,  384, 696,  384, 724,  356, 1778,  382, 724,  356, 696,  382, 1776,  384, 724,  356, 1804,  356, 696,  384, 724,  356, 1776,  384, 724,  356, 722,  358};  // DAIKIN176
uint8_t state[22] = {0x11, 0xDA, 0x17, 0x18, 0x04, 0x00, 0x1E, 0x11, 0xDA, 0x17, 0x18, 0x00, 0x73, 0x00, 0x21, 0x00, 0x00, 0x22, 0x35, 0x00, 0x20, 0x25};


Please advise.
Thank you and best regards

pasna committed:
Add files via upload

pasna committed:
add keywords

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -2337,10 +2342,12 @@ uint8_t IRDaikin176::convertFan(const stdAc::fanspeed_t speed) {
-    break;
-     default: setSwingHorizontal(kDaikin176SwingHAuto);
+    case kDaikin176SwingHAuto:
+      remote_state[kDaikin176ByteSwingH] &= kDaikin176MaskSwingH;
```
`~kDaikin176MaskSwingH`

crankyoldgit committed:
Daikin176: Fix Fan/SwingH problem.
Fix bit math problem.

* Add some unit tests to confirm what was and wasn't working.

Repeat after me: Unit tests are good! Write more Unit Tests!

crankyoldgit commented:
@pasna I've committed a fix to your `master` branch, so do a `git pull` to get it.

See: https://github.com/crankyoldgit/IRremoteESP8266/pull/826/commits/3683b12a4e3960684cc3f5e680b874a478507bb8

Basically, the problem was `setSwingHorizontal()` was clobbering the other half of the bit when you tried to clear the Swing bits with the bitwise AND.

Let me know when you're happy with it all and it is functioning how you want and we can look at merging.

You probably want to hand the "auto" fan speed properly. And check if there is a quiet speed too.

crankyoldgit reviewed (commented):
Let me know when you think it's ready.

pasna commented:
Your latest commit is working well. I can control my A/C unit model FHC30NUV2S. This model fan setting only 2 level Min and Max and nothing else.  

When I change mode to Fan or Dry I cannot tell if it function correctly, but I notice that when I press A/C remote to change
1) from Cool to Fan it shows change in byte 12 from 0x73 to 0x63 and byte 14 from 0x21 to 0x70 and byte 17 (temperature setting from any number to 0x10 = 17ºC).
2) from Fan to Dry it shows change in byte 12 from 0x63 to 0x23 and byte 14 from 0x71 to 0x71 and byte 17 stays at 0x10. And finally when change from Dry to Cool again it change the transmit codes back to original setting before mode changes. 
Please advise how to improve these coding to get IRMQTTServer to transmit same coding to A/C remote.

Thank you and have a great weekend.

crankyoldgit commented:
Can you give me the `state[]` lines for those various changes in the Real remote captured with/by DumpV2 please?

pasna commented:
This text file show what happened when I press mode change from Cool to Fan to Dry and back to Cool from real remote.
Best regards.
[Daikin176-mode-change-state.txt](https://github.com/crankyoldgit/IRremoteESP8266/files/3413212/Daikin176-mode-change-state.txt)

crankyoldgit commented:
> This model fan setting only 2 level Min and Max and nothing else.

There is no intermediate/medium setting between the two? Have you tried it? i.e. The A/C unit might be more capable than the remote says it is.

crankyoldgit committed:
Daikin176: Set Fan/Dry mode temp to 17C
* Fix another bit masking bug in setMode().
* Add unit test coverage for temp, mode, power, and swing.

FYI @pasna

crankyoldgit commented:
@pasna FYI, I added support for locking the temp to 17C when in Fan/Dry mode and fixed another `~` bit-math bug you had.

You'll want to `git pull` again.

I hope that does what you were asking for.

pasna commented:
Hi @crankyoldgit 
tested commit 552d7c7, Temperature change as expected when reading from IRrecvDumpV2 but A/C not respond i.e. cannot turn A/C on or change any mode at all. I noticed that byte12 (mode Cool) was 0x73 now it is 0x70 and last byte checksum has different value from real remote.
I send text file showing what I read from IRrecvDumpV2 
[Daikin176-mode-change-state.txt](https://github.com/crankyoldgit/IRremoteESP8266/files/3414393/Daikin176-mode-change-state.txt)

crankyoldgit commented:
I think I found the problem. Your mode mask was incorrect. Fixed that. 
e.g. Should have been: `const uint8_t kDaikin176MaskMode = 0b01110000;`
However, doing some more unit tests based on the data in the new text file uncovered you've got some byte/bit changes that are not explained.

e.g. (All from your "real" remote)

Power: On, Mode: 7 (COOL), Temp: 25C, Fan: 3 (MAX), Swing (H): 5 (Auto)  11DA171804001E11DA17180073**0021**00002035002023

to
Power: On, Mode: 6 (FAN), Temp: 17C, Fan: 3 (MAX), Swing (H): 5 (Auto)   11DA171804001E11DA17180063**0401**00010350020E7
state[13] changes from **0x00** to **0x04**.
state[14] changes from **0x21** to **0x01**.

to
Power: On, Mode: 2 (DRY), Temp: 17C, Fan: 3 (MAX), Swing (H): 5 (Auto)   11DA171804001E11DA1718002304**71**00001035002017
state[14] changes from **0x01** to **0x71**

to
Power: On, Mode: 7 (COOL), Temp: 25C, Fan: 3 (MAX), Swing (H): 5 (Auto)  11DA171804001E11DA1718007304**21**00002035002027
(Note: This is different code from the previous cool 25c state at the start. It differs at `state[13]`)
state[14] changes from **0x71** to **0x21**

Your code only has the `set/getPower()/on()/off()` touching state[14], nothing else according to your code uses or touches it.

These should be explained. Otherwise we are not accurately reproducing the same code

I'll upload a patch for the mode mask in a sec. As well as the unit tests that lead me to the above discovery.

crankyoldgit committed:
Daikin176: Fix incorrect mode mask.
* Add unit tests to verify against known-good remote data and try to
  reconstuct the messages. Currently we can't. See DEBUG lines in ir_Daikin_test.cpp

FYI @pasna

pasna committed:
Update ir_Daikin.h

pasna committed:
update keyword.txt

pasna committed:
Merge branch 'fixed-bugDaikin176-for-IRMQTTServer-' into master

pasna committed:
Merge branch 'master' into master

pasna commented:
Good morning,

Yes, your are correct and I am sorry for not be exact on the code.
These hex code are all from real remote and when pressing mode change it changes many bytes.
I repeat again, these code from real remote.

power button from OFF to ON,
Power: On, Mode: 7 (COOL), Temp: 25C, Fan: 3 (MAX), Swing (H): 5 (Auto)
11DA171804001E11DA17180073 **00** **21** 00002035002023
state[13] stays 0x00

mode change from COOL to FAN
Power: On, Mode: 6 (FAN), Temp: 17C, Fan: 3 (MAX), Swing (H): 5 (Auto)
11DA171804001E11DA17180063 **04** **01** 000010350020E7
state[13] change from 0x00 to 0x04
state[14] change from 0x21 to 0x01

mode change from FAN to DRY
Power: On, Mode: 2 (DRY), Temp: 17C, Fan: 3 (MAX), Swing (H): 5 (Auto)
11DA171804001E11DA17180023 **04** **71** 00001035002017
state[13] stays at 0x04
state[14] change from 0x01 to 0x71

mode change from DRY to COOL
Power: On, Mode: 7 (COOL), Temp: 25C, Fan: 3 (MAX), Swing (H): 5 (Auto)  11DA171804001E11DA17180073 **04** **21** 00002035002027
state[13] stays at 0x04
state[14] change from 0x71 to 0x21

when press power button from ON to OFF,
Power: Off, Mode: 7 (COOL), Temp: 25C, Fan: 3 (MAX), Swing (H): 5 (Auto)
11DA171804001E11DA17180073 **00** **20** 00002035002022
state[13] change from 0x04 to 0x00

You are right.

Thank you.

pasna committed:
remove whitespace

crankyoldgit commented on test/ir_Daikin_test.cpp:
```diff
@@ -2299,6 +2299,233 @@ TEST(TestDaikin160Class, HumanReadable) {
+  ac.setTemp(25);  // DEBUG Remove! Not needed. Added to pass.
+  ac.setRaw(on_cool_25_max_auto_v2);  // DEBUG Remove! Not needed. Added to pass
+  EXPECT_STATE_EQ(on_cool_25_max_auto_v2, ac.getRaw(), kDaikin176Bits);
+
```
This doesn't compile anymore, I'm pretty sure you are missing the ending `}` for the above function.

e.g. Here is the error message:
```
 TEST(TestDecodeDaikin128, RealExample) {
 ^
In file included from ../lib/googletest/googletest/include/gtest/gtest.h:59:0,
                 from ir_Daikin_test.cpp:8:
ir_Daikin_test.cpp:2532:1: error: qualified-id in declaration before ‘=’ token
 TEST(TestDecodeDaikin128, RealExample) {
 ^
ir_Daikin_test.cpp:2532:1: error: qualified-id in declaration before ‘(’ token
 TEST(TestDecodeDaikin128, RealExample) {
 ^
ir_Daikin_test.cpp:2591:1: error: expected ‘}’ at end of input
 }
```

crankyoldgit replied:
I've added a commit to fix this. Please `git pull`

crankyoldgit commented:
> Yes, your are correct and I am sorry for not be exact on the code.
> These hex code are all from real remote and when pressing mode change it changes many bytes.
> I repeat again, these code from real remote.

Yes, I know they are. However, _your_ code fails to be able reproduce these same messages.

Those **bold** bytes are changing in the _real_ messages, but your code _does not_ make similar changes, like the real remote does.

You need to work out what those changes are and why they change in the real remote messages, what it is linked to, so we can do the same in the ESP code version. Otherwise we are not producing the same message.

For example, Notice how you said two of the codes were the same, but the actual messages were different.
i.e. two different versions of the cold + on message. They should be the same, yet they are not. Something on the remote changed. It could be the clock on the remote if it has one, or some temperature sensor. I don't know. Only you can work that out as you have the remote and can see the display etc. There is still stuff about the bits in the real remote message you don't understand enough yet to all you to make a reliable emulation of the remote and it's messages.

crankyoldgit committed:
Fix an error introduce during a merge.

pasna commented:
Thanks for the comments. I will work hard on the remote.

Best regards

pasna committed:
Update ir_Daikin_test.cpp

pasna committed:
Merge branch 'master' of https://github.com/pasna/IRremoteESP8266

pasna commented:
I called my local customer service, confirmed that Daikin AC model FHC30NUV2S that I installed has only 2 fan setting Min and Max and nothing in between not ever Auto. I attached pictures of remote control model BRC4C153, this remote has only few function but has complex states. I am stuck, not able to get through this complexity. Need your help.
![IMG_8003](https://user-images.githubusercontent.com/7246333/61688449-3ed9e600-ad4f-11e9-979d-aaa2fb32a9fc.jpg)
![IMG_8004](https://user-images.githubusercontent.com/7246333/61688450-3ed9e600-ad4f-11e9-94cb-dea4359b371f.jpg)
![IMG_8005](https://user-images.githubusercontent.com/7246333/61688451-3ed9e600-ad4f-11e9-8621-bab51ab1ca04.jpg)

crankyoldgit commented:
Some remotes also encode which button was pressed in side one of the bytes.
Does state[13] or state[14] correlate to the the physical button pressed in this or other cases perhaps?
e.g. could **00** be "power button" and **04** be "mode button". Try other buttons to see if they change and have consistent values too.

crankyoldgit committed:
Daikin176: General improvements.
* Use/enforce only two values for fan speed.
* Use the new common routines in `toString()`.
* Add the fan speed and swing to the default/reset state.
* Simplify some code.
* Update unit tests accordingly.

FYI @pasna

crankyoldgit commented:
@pasna I've made some more changes. Please `git pull` from your own github repo to include them.
The changes are basically to enforce the two fan speeds per request.
I've also updated the routines in `::toString()` to the new ones used by other protocols.
Plus some other minor changes.

Regarding `SwingHorizontal()`, you've called it Horizontal (Left<->Right), but the image on the remote indicates it's probably Vertical(Up<->Down). Are you sure it is Horizontal?

pasna commented:
This model FHC series are ceiling mount (photo attached) and they have 4 vents. I decided to call that horizontal. Please correct me if I am wrong. 
Thank you
![IMG_7981](https://user-images.githubusercontent.com/7246333/61713441-c68d1800-ad82-11e9-90f8-f20bf55a668e.jpg)

pasna commented:
Real remote unit works like this -
state[13] will change from 0x00 to 0x04 only when mode button is push. Then it will change back to 0x00 again if power button or temperature button or swing button is being push.
state[14] changes comply to mode change  
at power ON 
from COOL to FAN change from 0x21 to 0x01
from FAN to DRY change from 0x01 to 0x71
from DRY to COOL change from 0x71 to 0x21
at POWER OFF
from COOL to FAN change from 0x20 to 0x00
from FAN to DRY change from 0x00 to 0x70
from DRY to COOL change from 0x70 to 0x20
Thank you and best regards

crankyoldgit commented:
I'll take a look and see about coding something up to support that.

crankyoldgit commented:
Currently we have coded up an "Auto" mode. Is there one for this A/C? If there isn't. What should we map requests for "Auto" too? i.e. What should the default be? Cool/Heat/Dry or Fan? (actually, is there a Heat mode?)

crankyoldgit commented:
If there are those modes, what are their `state[14]` changes?

pasna commented:
> Currently we have coded up an "Auto" mode. Is there one for this A/C? If there isn't. What should we map requests for "Auto" too? i.e. What should the default be? Cool/Heat/Dry or Fan? (actually, is there a Heat mode?)

> 
This model does not equip with AUTO or HEAT mode, for my region default should be COOL mode.

Thanks

crankyoldgit committed:
Daikin176: Handle mode setting better.
* remove Auto and Heat modes. Not supported.
* Default to cool mode.
* Handle internal message formation for modes better.
* Update unit tests accordingly.
* Remove hacks to get existing unit tests to pass.

FYI @pasna

crankyoldgit commented:
@pasna I've push more changes to your personal master branch/repo. `git pull`from your repo and let me know how those changes go.
I think that's now covering everything.

crankyoldgit committed:
Daikin176: Better handle modes for common a/c stuff

crankyoldgit commented:
Another update. I saw potential future issue.

pasna commented:
Good morning @crankyoldgit,
Tested e0e8281 and it works without problem. I am binding IRMQTTServer to Home Assistant and Homekit, it respond great on Home Assistant while Homekit for HASSIO is not totally match the protocol. HASSIO only support Thermostat not heaterCooler and Homekit does not has FAN mode and Dry Mode, of which actually not use in this South East Asia region anyway. 
I come across another repository of homebridge-mqttthing recently and it support heaterCooler characteristic on Homekit which is kind of interesting. May have it a try.

crankyoldgit commented:
FYI, this PR has been included in the v2.6.4 release of the library.
